plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformFileUtils.kt

```diff
@@ -128,6 +133,67 @@ fun parseXmlDependenciesReport(pathToXmlDependency: Path): DependencyUpdatesRepo
     return report
 }
 
+fun validateSctMetadata(sctFile: File?): SqlMetadataValidationResult {
+    if (sctFile == null) {
+        return SqlMetadataValidationResult(false, message("codemodernizer.chat.message.validation.error.missing_sct_file"))
+    }
+    val fileContent = sctFile.readBytes().toString(Charsets.UTF_8)
+    val xmlDeserializer = XmlMapper(JacksonXmlModule())
+    var sctMetadata: Map<*, *>? = null
+    try {
+        sctMetadata = xmlDeserializer.readValue(fileContent, Any::class.java) as Map<*, *>
+    } catch (e: Exception) {
+        getLogger<CodeTransformChatController>().error { "Error parsing .sct metadata file; invalid XML encountered." }
+        return SqlMetadataValidationResult(false, message("codemodernizer.chat.message.validation.error.invalid_sct"))
+    }
+
+    try {
+        val instances = sctMetadata["instances"] as Map<*, *>
+        val projectModel = instances["ProjectModel"] as Map<*, *>
+        val entities = projectModel["entities"] as Map<*, *>
+
+        val sources = entities["sources"] as Map<*, *>
+        val sourceDbServer = sources["DbServer"] as Map<*, *>
+        val sourceVendor = (sourceDbServer["vendor"] as String).trim().uppercase()
+        if (sourceVendor != "ORACLE") {
+            return SqlMetadataValidationResult(false, message("codemodernizer.chat.message.validation.error.invalid_source_db"))
+        }
+
+        val sourceServerName = (sourceDbServer["name"] as String).trim()
+
+        val targets = entities["targets"] as Map<*, *>
+        val targetDbServer = targets["DbServer"] as Map<*, *>
+        val targetVendor = (targetDbServer["vendor"] as String).trim().uppercase()
```

Comment:
treating the contents as an arbitrary bag of maps defeats the point of type safety